ingester/src/query/projection.rs

```
@@ -0,0 +1,129 @@
+                }
+            }
+
+            Projection::Project(cols) => {
+                // Invariant: subset is never empty
+                assert!(!cols.is_empty());
+
+                // Construct the schema & data arrays in a single pass, ordered
+                // by the projection and ignoring any missing columns.
+                for name in cols {
+                    if let Ok(column) = batch.column(name) {
+                        schema_builder.influx_column(name, column.influx_type());
+                        column_data
+                            .push(column.to_arrow().expect("failed to snapshot buffer data"));
+                    }
+                }
+            }
+        };
+
+        let schema = schema_builder
+            .build()
+            .expect("failed to create batch schema");
+
+        RecordBatch::try_new(schema.into(), column_data)
+            .expect("failed to generate snapshot record batch")
+    }
+
+    /// Apply the specified projection to `batches`.
+    ///
+    /// This projection requires relatively cheap ref-counting clones and does
+    /// not copy the underlying data.
+    pub(crate) fn project_record_batch(&self, batches: &[RecordBatch]) -> Vec<RecordBatch> {
+        match &self.0 {
+            Projection::All => batches.to_vec(),
+            Projection::Project(columns) => {
+                // Invariant: subset is never empty
+                assert!(!columns.is_empty());
+
+                batches
+                    .iter()
+                    .map(|batch| {
+                        let schema = batch.schema();
+
+                        // Map the column names to column indexes, ignoring
+                        // columns specified in the columns that do not exist
+                        // in this batch.
+                        let projection = columns
+                            .iter()
+                            .flat_map(|column_name| schema.index_of(column_name).ok())
```

BTW: schema is NOT a Map but a Vec internally. For many columns, this lookup is $O(n^2)$ complexity (which for 200+ cols might be an issue). I usually build a lookup map before performing projections (and even cache the map in many cases). Not saying you HAVE to do that here, just want to point that out.

Somewhat related: if that helps you, the querier could even provide you the projection as column IDs instead of strings. Open a ticket if you want to have that and I make that possible.

---

We don't have column IDs, so I don't think it'll help - it also wouldn't work for not-yet-persisted data, as it would force the ingester to do the lookup queries on the hot path :( 

I copy-pasta'd this code from the existing impl somewhere - is there a better way of doing this today?

---

> I copy-pasta'd this code from the existing impl somewhere - is there a better way of doing this today?

Not without additional caching / complexity.

---

Is the reason to use `ok()` here (and ignore the error) if the particular ingester partition doesn't have one or more of the columns?